File: src/main/kotlin/cn/solarmoon/sparkcore/api/geometry/model/Cube.kt
package cn.solarmoon.sparkcore.api.geometry.model

import cn.solarmoon.sparkcore.api.util.Direction
import cn.solarmoon.sparkcore.api.math.Float2
import cn.solarmoon.sparkcore.api.math.Float3
import cn.solarmoon.sparkcore.api.math.Mat4
import cn.solarmoon.sparkcore.api.math.Quaternionf
import cn.solarmoon.sparkcore.api.math.cartesianProduct
import cn.solarmoon.sparkcore.api.math.rotate
import cn.solarmoon.sparkcore.api.math.rotateLocalZYX
import cn.solarmoon.sparkcore.api.math.toFloat
import cn.solarmoon.sparkcore.api.math.translate
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.Transient

/**
 * 顶点顺序规范 (Corners):
 *
 * ```
 *             topLeftBack (2) ----------- topRightBack (3)
 *                /|                          /|
 *               / |                         / |
 *              /  |                        /  |
 *   topLeftFront (4) ------ topRightFront (5) |
 *             |   |                       |   |
 *             |   |                       |   |
 *             |   |                       |   |
 *             |  bottomLeftBack (0) ------|-- bottomRightBack (1)
 *             |  /                        |  /
 *             | /                         | /
 *             |/                          |/
 *   bottomLeftFront (6) -------- bottomRightFront (7)
 * ```
 *
 * 八个角点：
 *   (0) bottomLeftBack
 *   (1) bottomRightBack
 *   (2) topLeftBack
 *   (3) topRightBack
 *   (4) topLeftFront
 *   (5) topRightFront
 *   (6) bottomLeftFront
 *   (7) bottomRightFront
 *
 * 十二条边：
 *   底面: (0–1), (1–7), (7–6), (6–0)
 *   顶面: (2–3), (3–5), (5–4), (4–2)
 *   竖直: (0–2), (1–3), (7–5), (6–4)
 */
@Serializable
data class Cube(
    val origin: Float3,
    val pivot: Float3 = Float3(),
    val rotation: Float3 = Float3(),
    val size: Float3,
    val inflate: Float = 0f,
    @SerialName("uv") val uvUnion: UVUnion,
    val mirror: Boolean = false,
) {

    @Transient
    val rotationQuaternion = Quaternionf().rotateLocalZYX(rotation)

    @Transient
    val transform = Mat4()
        .translate(origin)
        .translate(pivot)
        .rotate(rotationQuaternion)
        .translate(-pivot)

    internal val corners by lazy {
        val xs = listOf(origin.x - inflate, origin.x + size.x + inflate)
        val ys = listOf(origin.y - inflate, origin.y + size.y + inflate)
        val zs = listOf(origin.z - inflate, origin.z + size.z + inflate)
        val coords = cartesianProduct(listOf(xs, ys, zs)).map { (x, y, z) ->
            Float3(x, y, z)
        }
        Corners(
            coords[0],
            coords[4],
            coords[2],
            coords[6],
            coords[3],
            coords[7],
            coords[1],
            coords[5]
        )
    }

    internal data class Corners(
        val bottomLeftBack: Float3,   // 0
        val bottomRightBack: Float3,  // 1
        val topLeftBack: Float3,      // 2
        val topRightBack: Float3,     // 3
        val topLeftFront: Float3,     // 4
        val topRightFront: Float3,    // 5
        val bottomLeftFront: Float3,  // 6
        val bottomRightFront: Float3  // 7
    ) {
        // 左下 → 右下 → 右上 → 左上
        val quadWest  get() = listOf(bottomLeftBack, bottomRightBack, topRightBack, topLeftBack)
        val quadEast  get() = listOf(bottomRightBack, bottomRightFront, topRightFront, topRightBack)
        val quadNorth get() = listOf(bottomLeftBack, bottomLeftFront, topLeftFront, topLeftBack)
        val quadSouth get() = listOf(bottomRightBack, bottomRightFront, topRightFront, topRightBack)
        val quadUp    get() = listOf(topLeftBack, topRightBack, topRightFront, topLeftFront)
        val quadDown  get() = listOf(bottomLeftBack, bottomRightBack, bottomRightFront, bottomLeftFront)
    }

}

fun Cube.getQuadPositions(direction: Direction) = with(corners) {
    when (direction) {
        Direction.WEST -> if (mirror) quadEast else quadWest
        Direction.EAST -> if (mirror) quadWest else quadEast
        Direction.NORTH -> quadNorth
        Direction.SOUTH -> quadSouth
        Direction.UP -> if (mirror && uvUnion is UVUnion.Face) quadDown else quadUp
        Direction.DOWN -> if (mirror && uvUnion is UVUnion.Face) quadUp else quadDown
    }
}

fun Cube.getQuadVertices(direction: Direction, meta: ModelDescription): List<Vertex> {
    val poses = getQuadPositions(direction)
    val normal = direction.normal.toFloat()
    return when(uvUnion) {
        is UVUnion.Box -> {
            val sx = size.x
            val sy = size.y
            val sz = size.z
            val ux = uvUnion.uv.x
            val uy = uvUnion.uv.y

            // 按BB矩形展开规范获取每个面的矩形区域（左上起点+矩形长宽）
            val (u, v, w, h) = when (direction) {
                Direction.EAST  -> listOf(ux, uy + sz, sz, sy)
                Direction.WEST  -> listOf(ux + sx + sz, uy + sz, sz, sy)
                Direction.NORTH -> listOf(ux + sz, uy + sz, sx, sy)
                Direction.SOUTH -> listOf(ux + sx + sz + sz, uy + sz, sx, sy)
                Direction.UP    -> listOf(ux + sz, uy, sx, sz)
                Direction.DOWN  -> listOf(ux + sz + sx, uy, sx, sz)
            }

            // 计算矩形四边
            val u1o = u
            val u2o = u + w
            val v1o = v
            val v2o = v + h

            // 应用 mirror
            val (u1, u2) = if (mirror) u2o to u1o else u1o to u2o
            val (v1, v2) = v1o to v2o

            val uvList = listOf(
                Float2(u1, v1),
                Float2(u2, v1),
                Float2(u2, v2),
                Float2(u1, v2)
            )

            poses.zip(uvList) { pos, uv ->
                Vertex(pos, uv, normal)
            }
        }
        is UVUnion.Face -> {
            val faceUV = uvUnion.uvMap[direction]!!
            val u1o = faceUV.uv.x / meta.textureWidth
            val v1o = faceUV.uv.y / meta.textureHeight
            val u2o = (faceUV.uv.x + faceUV.uvSize.x) / meta.textureWidth
            val v2o = (faceUV.uv.y + faceUV.uvSize.y) / meta.textureHeight

            // 应用 mirror
            val (u1, u2) = if (mirror) u2o to u1o else u1o to u2o
            val (v1, v2) = v1o to v2o

            val uvList = listOf(
                Float2(u1, v1),
                Float2(u2, v1),
                Float2(u2, v2),
                Float2(u1, v2)
            )

            poses.zip(uvList) { pos, uv ->
                Vertex(pos, uv, normal)
            }
        }
    }
}

File: rust/crates/spark-core-geometry/src/model/cube.rs
use crate::model::model_description::ModelDescription;
use crate::model::uv::{Direction, UVUnion};
use crate::model::vertex::Vertex;
use glam::{Mat4, Quat, Vec2, Vec3};
#[cfg(feature = "serde")]
use serde::{Deserialize, Serialize};

#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[derive(Debug, Clone, Copy)]
pub struct Cube {
    #[cfg_attr(feature = "serde", serde(default))]
    pub origin: Vec3,
    #[cfg_attr(feature = "serde", serde(default))]
    pub pivot: Vec3,
    #[cfg_attr(feature = "serde", serde(default))]
    pub rotation: Quat,
    #[cfg_attr(feature = "serde", serde(default))]
    pub size: Vec3,
    #[cfg_attr(feature = "serde", serde(default))]
    pub inflate: f32,
    #[cfg_attr(feature = "serde", serde(default))]
    pub mirror: bool,
    pub uv: UVUnion
}

impl Cube {
    pub fn local_transform(&self) -> Mat4 {
        Mat4::from_translation(self.origin)
            * Mat4::from_translation(self.pivot)
            * Mat4::from_quat(self.rotation)
            * Mat4::from_translation(-self.pivot)
    }

    /// 获取指定面的四个原始物理坐标
    fn quad_positions(&self, direction: Direction) -> [Vec3; 4] {
        let min = self.origin - Vec3::splat(self.inflate);
        let max = self.origin + self.size + Vec3::splat(self.inflate);
        let mut quad = match direction {
            Direction::North => [
                Vec3::new(max.x, min.y, min.z), Vec3::new(min.x, min.y, min.z),
                Vec3::new(min.x, max.y, min.z), Vec3::new(max.x, max.y, min.z),
            ],
            Direction::South => [
                Vec3::new(min.x, min.y, max.z), Vec3::new(max.x, min.y, max.z),
                Vec3::new(max.x, max.y, max.z), Vec3::new(min.x, max.y, max.z),
            ],
            Direction::West => [
                Vec3::new(min.x, min.y, min.z), Vec3::new(min.x, min.y, max.z),
                Vec3::new(min.x, max.y, max.z), Vec3::new(min.x, max.y, min.z),
            ],
            Direction::East => [
                Vec3::new(max.x, min.y, max.z), Vec3::new(max.x, min.y, min.z),
                Vec3::new(max.x, max.y, min.z), Vec3::new(max.x, max.y, max.z),
            ],
            Direction::Up => [
                Vec3::new(max.x, max.y, max.z), Vec3::new(max.x, max.y, min.z),
                Vec3::new(min.x, max.y, min.z), Vec3::new(min.x, max.y, max.z),
            ],
            Direction::Down => [
                Vec3::new(min.x, min.y, min.z), Vec3::new(max.x, min.y, min.z),
                Vec3::new(max.x, min.y, max.z), Vec3::new(min.x, min.y, max.z),
            ],
        };

        // 处理镜像互换
        if self.mirror {
            match direction {
                Direction::West | Direction::East => quad.reverse(),
                Direction::Up | Direction::Down if matches!(self.uv, UVUnion::Face { .. }) => quad.reverse(),
                _ => {}
            }
        }
        quad
    }

    /// 生成一个面的四个顶点
    pub fn quad_vertices(&self, direction: Direction, meta: &ModelDescription) -> [Vertex; 4] {
        let texture_width = meta.texture_width as f32;
        let texture_height = meta.texture_height as f32;

        let transform = self.local_transform();
        let mut positions = self.quad_positions(direction);
        let normal = direction.normal();

        for pos in &mut positions {
            *pos = transform.transform_point3(*pos);
        }

        let (u1o, v1o, u2o, v2o) = match &self.uv {
            UVUnion::Box(uv) => {
                let (sx, sy, sz) = (self.size.x, self.size.y, self.size.z);
                let (ux, uy) = (uv.x, uv.y);

                let (u, v, w, h) = match direction {
                    Direction::East  => (ux, uy + sz, sz, sy),
                    Direction::West  => (ux + sx + sz, uy + sz, sz, sy),
                    Direction::North => (ux + sz, uy + sz, sx, sy),
                    Direction::South => (ux + sx + sz + sz, uy + sz, sx, sy),
                    Direction::Up    => (ux + sz, uy, sx, sz),
                    Direction::Down  => (ux + sz + sx, uy, sx, sz),
                };

                (
                    u / texture_width,
                    v / texture_height,
                    (u + w) / texture_width,
                    (v + h) / texture_height
                )
            }

            UVUnion::Face(uv_map) => {
                let face_uv = uv_map[direction];
                (
                    face_uv.uv.x / texture_width,
                    face_uv.uv.y / texture_height,
                    (face_uv.uv.x + face_uv.uv_size.x) / texture_width,
                    (face_uv.uv.y + face_uv.uv_size.y) / texture_height,
                )
            }
        };

        let (u1, u2) = if self.mirror { (u2o, u1o) } else { (u1o, u2o) };
        let (v1, v2) = (v1o, v2o);

        let uvs = [
            Vec2::new(u1, v1),
            Vec2::new(u2, v1),
            Vec2::new(u2, v2),
            Vec2::new(u1, v2),
        ];

        [
            Vertex { position: positions[0], uv: uvs[0], normal },
            Vertex { position: positions[1], uv: uvs[1], normal },
            Vertex { position: positions[2], uv: uvs[2], normal },
            Vertex { position: positions[3], uv: uvs[3], normal },
        ]
    }

    pub fn all_vertices(&self, meta: &ModelDescription) -> [Vertex; 24] {
        let n = self.quad_vertices(Direction::North, meta);
        let s = self.quad_vertices(Direction::South, meta);
        let e = self.quad_vertices(Direction::East, meta);
        let w = self.quad_vertices(Direction::West, meta);
        let u = self.quad_vertices(Direction::Up, meta);
        let d = self.quad_vertices(Direction::Down, meta);

        [
            n[0], n[1], n[2], n[3],
            s[0], s[1], s[2], s[3],
            e[0], e[1], e[2], e[3],
            w[0], w[1], w[2], w[3],
            u[0], u[1], u[2], u[3],
            d[0], d[1], d[2], d[3],
        ]
    }

    pub fn bake(self, meta: &ModelDescription) -> BakedCubeData {
        let vertices = self.all_vertices(meta);
        BakedCubeData {
            cube: self,
            vertices
        }
    }
}

#[derive(Debug, Clone)]
pub struct BakedCubeData {
    pub cube: Cube,
    pub vertices: [Vertex; 24],
}